pkg/server/middleware/webhook.go

```diff
@@ -2,50 +2,49 @@ package middleware
 
 import (
 	"bytes"
-	"context"
 	"net/http"
 
-	"github.com/tbd54566975/ssi-service/pkg/server/framework"
+	"github.com/gin-gonic/gin"
+	"github.com/pkg/errors"
+
 	svcframework "github.com/tbd54566975/ssi-service/pkg/service/framework"
 	"github.com/tbd54566975/ssi-service/pkg/service/webhook"
 )
 
 // Define a custom response writer that wraps the original response writer and writes to a buffer
 type responseWriter struct {
-	http.ResponseWriter
+	gin.ResponseWriter
 	buf *bytes.Buffer
 }
 
 func (rw *responseWriter) Write(b []byte) (int, error) {
 	// Write the response body to both the original response writer and the buffer
-	n, err := rw.ResponseWriter.Write(b)
-	if err != nil {
+	if n, err := rw.buf.Write(b); err != nil {
 		return n, err
 	}
-
-	_, err = rw.buf.Write(b)
-	if err != nil {
-		return n, err
-	}
-
-	return n, err
+	return rw.ResponseWriter.Write(b)
 }
 
-func Webhook(webhookService svcframework.Service, noun webhook.Noun, verb webhook.Verb) framework.Middleware {
-	return func(handler framework.Handler) framework.Handler {
-		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
-			buf := bytes.NewBuffer([]byte{})
-
-			// Wrap the original response writer with a new response writer that writes to the buffer
-			wrappedWriter := &responseWriter{w, buf}
-
-			// Call the original handler with the new response writer
-			err := handler(ctx, wrappedWriter, r)
-
-			whService := webhookService.(*webhook.Service)
-			go whService.PublishWebhook(noun, verb, buf)
-
-			return err
+// Webhook is a middleware that publishes a webhook after the request handler has finished writing the response
+// TODO(https://github.com/TBD54566975/ssi-service/issues/435): currently this runs on each request even if no webhooks are registered. It should be updated to only run if webhooks are registered.
+func Webhook(webhookService svcframework.Service, noun webhook.Noun, verb webhook.Verb) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		// Wrap the original response writer with a new response writer that writes to the buffer
+		buf := bytes.NewBuffer([]byte{})
+		wrappedWriter := &responseWriter{ResponseWriter: c.Writer, buf: buf}
+		c.Writer = wrappedWriter
+
+		// run the middleware after the request handler
+		c.Next()
+
+		// make sure the response was already written
+		if !c.Writer.Written() {
```

Comment:
what if there is nothing in the buffer, like a delete, it just returns nil, will the error here?

webhook should still be able to fire on a delete with a nil response

Reply:
good point. I will change this to a warning log